File: kartwheel/app/src/main/java/us/handstand/kartwheel/model/Point.java
package us.handstand.kartwheel.model;


import android.support.annotation.NonNull;

import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;

@AutoValue
public abstract class Point implements PointModel, Comparable<Point> {
    @Override
    public int compareTo(@NonNull Point o) {
        return stepOrder().compareTo(o.stepOrder());
    }

    // Required by Gson
    @SuppressWarnings("unused")
    public static TypeAdapter<Point> typeAdapter(Gson gson) {
        return new AutoValue_Point.GsonTypeAdapter(gson);
    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/model/Point.kt
package us.handstand.kartwheel.model


import com.google.auto.value.AutoValue
import com.google.gson.Gson
import com.google.gson.TypeAdapter
import us.handstand.kartwheel.model.PointModel.Creator

@AutoValue
abstract class Point : PointModel, Comparable<Point> {
    override fun compareTo(other: Point): Int = stepOrder()!!.compareTo(other.stepOrder()!!)

    companion object : Creator<Point> by Creator(::AutoValue_Point) {
        // Required by Gson
        @JvmStatic
        fun typeAdapter(gson: Gson): TypeAdapter<Point> = AutoValue_Point.GsonTypeAdapter(gson)
    }
}
